onnx/optimizer/passes/fuse_bn_into_conv.h

```diff
@@ -0,0 +1,212 @@
+    }
+
+    if (conv->inputs().size() == 3) {
+      Value* new_b_value = graph.addInitializerAndInput(b);
+      Value* old_b_value = conv->inputs()[2];
+      conv->replaceInput(2, new_b_value);
+      if (old_b_value->uses().size() == 0) {
+        graph.eraseInitializerAndInput(old_b_value);
+      }
+    } else {
+      Value* new_b_value = graph.addInitializerAndInput(b);
+      conv->addInput(new_b_value);
+    }
+  }
+
+  bool modify_conv(Node* conv, Node* bn, Graph& graph)  {
+    const auto& bn_inputs = bn->inputs();
+    const auto& conv_inputs = conv->inputs();
+    auto end_iter = graph.initializers().end();
+    auto s_iter = graph.getInitializer(bn_inputs[1]->uniqueName());
+    auto bbn_iter = graph.getInitializer(bn_inputs[2]->uniqueName());
+    auto m_iter = graph.getInitializer(bn_inputs[3]->uniqueName());
+    auto var_iter = graph.getInitializer(bn_inputs[4]->uniqueName());
+    auto W_iter = graph.getInitializer(conv_inputs[1]->uniqueName());
+    if (s_iter == end_iter || bbn_iter == end_iter || m_iter == end_iter ||
+        var_iter == end_iter || W_iter == end_iter) {
+      return false;
+    }
+
+    ONNX_ASSERT(s_iter->sizes().size() == 1);
+    ONNX_ASSERT(bbn_iter->sizes().size() == 1 && bbn_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(m_iter->sizes().size() == 1 && m_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(var_iter->sizes().size() == 1 && var_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(W_iter->sizes().size() > 2 && W_iter->sizes()[0] == s_iter->sizes()[0]);
+    ONNX_ASSERT(s_iter->elem_type() == bbn_iter->elem_type() &&
+                s_iter->elem_type() == m_iter->elem_type() &&
+                s_iter->elem_type() == var_iter->elem_type() &&
+                s_iter->elem_type() == W_iter->elem_type());
+    if (s_iter->elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT &&
+        s_iter->elem_type() != ONNX_NAMESPACE::TensorProto_DataType_DOUBLE)  {
+      return false;
+    }
+
+    auto s = *s_iter;
+    auto bbn = *bbn_iter;
+    auto m = *m_iter;
+    auto var = *var_iter;
+    auto W = *W_iter;
+    auto epsilon = bn->hasAttribute(kepsilon) ? bn->f(kepsilon) : 1e-5;
+    Tensor eps;
+    Tensor bc;
+
+    switch(s.elem_type()) {
+      case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
+        eps.sizes().push_back(s.sizes()[0]);
+        eps.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+        for (int64_t i = 0; i < eps.sizes()[0]; ++i)  {
+          eps.floats().push_back((float) epsilon);
+        }
+        bc.sizes().push_back(s.sizes()[0]);
+        bc.elem_type() = ONNX_NAMESPACE::TensorProto_DataType_FLOAT;
+        for (int64_t i = 0; i < eps.sizes()[0]; ++i)  {
+          bc.floats().push_back(0.f);
+        }
+
+        if (conv_inputs.size() == 3) {
+          auto bc_iter = graph.getInitializer(conv_inputs[2]->uniqueName());
+          if (bc_iter == end_iter) {
+            return false;
```

Comment:
it's possible the original conv doesn't bias but the fused one has, so here should add a new bias to the fused conv instead of stop doing optimization

Reply:
This code is not stopping the optimization if the original conv doesn't have bias. It is checking if there are 3 inputs (i.e. input of a bias). If so, we get the initializer; if that input is not an initializer, then we cannot do the optimization as we do not know the value. If there aren't 3 inputs, we still have a bc tensor - defined above - and we replace it in the replace_inputs function

Reply:
In that case could you move this bias check to above (where you checked whether other inputs are initializers).

Reply:
moved